module/jobqueue/README.md

```diff
@@ -0,0 +1,70 @@
+## JobQueue Design Goal
+
+The jobqueue package implemented a reusable job queue system for async message processing.
+
+The most common use case is to work on each finalized block async.
+
+For instance, verification nodes must verify each finalized block. This needs to happen async, otherwise a verification node might get overwhelmed during periods when a large amount of blocks are finalized quickly (e.g. when a node comes back online and is catching up from other peers).
+
+So the goal for the jobqueue system are:
+1. guarantee each job (i.e. finalized block) will be processed eventually
+2. in the event of a crash failure, the jobqueue state is persisted and workers can be rescheduled so that no job is skipped.
+3. allow concurrent processing of multiple jobs
+4. the number of concurrent workers is configurable so that the node won't get overwhelmed when too many jobs are created (i.e. too many blocks are finalized in a short period of time)
+
+## JobQueue components
+To achieve the above goal, the jobqueue system contains the following components/interfaces:
+1. A `Jobs` module to find jobs by job index
+2. A `storage.ConsumerProgress` to store job processing progress
+3. A `Worker` module to process jobs and report job completion.
+4. A `Consumer` that orchestrates the job processing by finding new jobs, creating workers for each job using the above modules, and managing job processing status internally.
+
+### Using module.Jobs to find jobs
+There is no JobProducer in jobqueue design. Job queue assumes each job can be indexed by a uint64 value, just like each finalized block (or sealed block) can be indexed by block height.
+
+Let's just call this uint64 value "Job Index" or index.
+
+So if we iterate through each index from low to high, and find each job by index, then we are able to iterate through each job.
+
+Therefore modules.Job interface abstracts it into a method: `AtIndex`.
+
+`AtIndex` method returns the job at any given index.
+
+Job consumer relies on the modules.Jobs to find jobs. However, modules.Jobs doesn't provide a way to notify as soon as a new job is available. So it's consumer's job to keep track of the values returned by module.Jobs's `Head` method and find jobs that are new.
+
+### Using Check method to notify job consumer for checking new jobs
+Job consumer provides the `Check` method for users to notify new jobs available.
+
+Once called, job consumer will iterate through each height with the `AtIndex` method. It stops when one of the following condition is true:
+1. no job was found at a index
+2. no more workers to work on them
+3. max search ahead is reached (will explain later)
```

Comment:
Not sure what does max search ahead means here.  Looking at the code `maxProcessing` is the only limit there (which is the same as max number of workers).

Reply:
Good catch.

The max search ahead [is implemented on a different branch,](https://github.com/onflow/flow-go/pull/1895/files#diff-2731207c7c63d225021f36118b9379bcb3e4394d9ef95e314020a55ff0715500R33) but not merged to master yet.

I added it to the TODOs, which are the features extended by @peterargue ['s PR. ](https://github.com/onflow/flow-go/pull/1895/files#diff-8eb19c06c1184bbe002ad8b1ccaf04394e602c64077c96943b2947757fb92c63)